### ISSUE TITLE
Unable to get any statistics. ERROR: Cannot parse LoxPLAN XML

### ISSUE THREAD

**Issue by ActiveByte** [mschlenstedt/LoxBerry-Plugin-Stats4Lox-NG#131]
When i go to the 'Loxone and import' page i don't get any results after the page has loaded.
When i go to the logs i see the following error:

`CRITICAL: import.cgi: Cannot parse LoxPLAN XML file: :3: parser error : error parsing attribute name
rea="1" UnitHeight="1" UnitWindSpeed="4" UnitPrecipitation="2" UnitPressure="2"
^
:3: parser error : attributes construct error
rea="1" UnitHeight="1" UnitWindSpeed="4" UnitPrecipitation="2" UnitPressure="2"
^
:3: parser error : Couldn't find end of Start Tag C line 3
rea="1" UnitHeight="1" UnitWindSpeed="4" UnitPrecipitation="2" UnitPressure="2"
^
:3180: parser error : Opening and ending tag mismatch: ControlList line 2 and C

^
:3181: parser error : Extra content at the end of the document

^

 CRITICAL: Error parsing XML`

**Comment by ActiveByte** [comment]
I was able to solve this issue by changing my units in Loxone config under the main configuration item in the list.

**Comment by christianTF** [comment]
Do you still have the uncorrected file for testing?
If you send me that file via PM on loxforum.com, or by email christian at loxberry.de, I possibly can automatically fix the xml before handing it over to the xml parser (we do that already for several Loxone non-valid xml parts in our code).

**Comment by smiley125** [comment]
I'm having the same issue, with the same error, just wondering how you resolved it?

My error is:

`CRITICAL: import.cgi: Cannot parse LoxPLAN XML file: :3: parser error : error parsing attribute name
6b2b-ffffa3b5cadb2c5f" UnitWindSpeed="4" UnitPrecipitation="1" UnitPressure="1"
^
:3: parser error : attributes construct error
6b2b-ffffa3b5cadb2c5f" UnitWindSpeed="4" UnitPrecipitation="1" UnitPressure="1"
^
:3: parser error : Couldn't find end of Start Tag C line 3
6b2b-ffffa3b5cadb2c5f" UnitWindSpeed="4" UnitPrecipitation="1" UnitPressure="1"
^
:4805: parser error : Opening and ending tag mismatch: ControlList line 0 and C

^
:4806: parser error : Extra content at the end of the document

^`

**Comment by MephistoJB** [comment]
I think there is another error with parsing LoxPLAN XML

 CRITICAL: import.cgi: Cannot parse LoxPLAN XML file: :335: parser error : Attribute Phn redefined
UsRnA="1664" iBtnArr="" CodeArr="" NFCArr="" NFCNArr="" EAIdArr="" LastEdit="0"
^
:342: parser error : Attribute Phn redefined
UsRnA="1664" iBtnArr="" CodeArr="" NFCArr="" NFCNArr="" EAIdArr="" LastEdit="0"
^
:346: parser error : Attribute Phn redefined
UsRnA="1024" iBtnArr="" CodeArr="" NFCArr="" NFCNArr="" EAIdArr="" LastEdit="0"
^
:350: parser error : Attribute Phn redefined
1664" iBtnArr="" CodeArr="" NFCArr="" NFCNArr="" EAIdArr="" LastEdit="418248774"
^

 CRITICAL: Error parsing XML

**Comment by MephistoJB** [comment]
Maybe also worth mentioning that I use Loxone Config Version 14.3.8.3 and a Miniserver Gen. 1

**Comment by mr-manuel** [comment]
Is this maybe something similar to https://github.com/mschlenstedt/LoxBerry-Plugin-Stats4Lox-NG/issues/139? I'm also using a Miniserver Gen. 1 and was using 14.3.8.3, now 14.4.8.9, but no issues with that.
If you search that lines in the *.Loxone file, what is the whole block?

**Comment by MephistoJB** [comment]
yes, this is exactly the same. I followed the guidance in the forum and it fixed it. But I had to do this manually. Does this come in the next version?

**Comment by MephistoJB** [comment]
You sure? @christianTF reopened it on purpose in January

**Comment by smiley125** [comment]
> yes, this is exactly the same. I followed the guidance in the forum and it fixed it. But I had to do this manually. Does this come in the next version?
> 
> Oh and thanks

Can you link the guidance? Still struggling with mine.

Thanks

**Comment by MephistoJB** [comment]
https://www.loxforum.com/forum/projektforen/loxberry/plugins/315626-stats4lox-das-statistik-plugin?p=400051#post400051

But its german :)

**Comment by christianTF** [comment]
Michael has commited another known issue in Loxone‘s XML, please update the plugin to the current master. This may fix it.
If not, we need the full line of the Loxone XML that fails, as the error message in the log does not display the block type where it happens.

**Comment by mschlenstedt** [comment]
Fixed in [`f448274`](https://github.com/mschlenstedt/LoxBerry-Plugin-Stats4Lox-NG/commit/f448274).

The parser is no longer all-or-nothing. Loading the LoxPLAN now happens in three logged stages: strict, then duplicate attributes removed from **all** element types (not just three hard coded ones), then libxml2 recover mode. So a single invalid element does not abort the whole import any more.

The other half of this issue was the diagnosis. libxml2 only reports a line number, which told neither you nor us which block is broken - that is why reports like the one here could never be concluded. The element is now resolved from the source:

```
LoxPLAN line 7209: Unescaped '<' not allowed in attributes values
   affected block: Type="Memory" Title="<&- Status EG"
   source line   : <C Type="Memory" Desc="1-Wire EG" ... >
```

If even recover mode fails, a copy of the file is kept as `data/plugins/stats4lox/loxplan_parse_failed.xml` so it can be attached to a report.

Verified against a real 2.9 MB LoxPLAN with 4656 elements, including a deliberately corrupted attribute value of the kind reported here - 10 of 10 checks pass. Please reopen if a file still fails after this.
